enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,182 @@
+---
+title: neat-enhancement-idea
+authors:
+  - "@Danil-Grigorev"
+reviewers:
+  - "@enxebre"
+  - "@elmiko"
+  - "@mfedosin"
+approvers:
+  - "@enxebre"
+  - "@elmiko"
+  - TBD
+creation-date: 2020-08-31
+last-updated: 2020-09-01
+status: implementable
+see-also:
+  - "/enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md"  
+replaces:
+superseded-by:
+---
+
+# Add support for out-of-tree cloud provider integration
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+1) Should we running `cloud-controller-manager` as:
+- static `Pod` managed from the `Node` filesystem.
+- `Deployment` backing up the `Pod`.
+- `ReplicaSet` to run a `Pod` on each master node.
+
+## Summary
+
+This enhancement proposal describes steps needed to add support for out-of-tree providers in cloud providers currently supported by Openshift. This includes: AWS, Azure, GCP, vSphere, IBM. Openstack provider proposal is located at it's dedicated document.
+
+## Motivation
+
+Upstream is currently moving towards exclusion of in-tree cloud-provider code from `k/k` repository, and instead introducing out-of-tree provider support, making them independent from `k/k` releases, and simplifying the process of development and e2e testing for those plugins. Previously each cloud-provider was located in the [`legacy-cloud-providers`](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/legacy-cloud-providers) and now the support and main development is done in the dedicated cloud-provider [repositories](https://github.com/kubernetes?q=cloud-provider-). Upstream is planning to remove in-tree support completely in the future releases, approximetly in [v1.21](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600).
+
+### Goals
+
+- Prepare openshift components to accomodate the out-of-tree plugins and remove support for in-tree implementation.
+- Add an installer option to select an`external` cloud provider.
+- Deploy cluster infrastructure accordingly to selected `external` option for cloud provider, and allow `cloud-controller` component registration via the provider specific upstream implemetation.
+
+### Non-Goals
+
+- Force immediate exclusion of in-tree support for currently used providers, as their out-of-tree counterparts are added.
```

Comment:
Does this imply there would be some period where OpenShift would support both?  That doesn't seem ideal ...

As @JoelSpeed noted in another comment, "Surely a user doesn't need to know about this implementation detail?"

Reply:
If it is done correctly, we could end up with a situation when one provider will be using external configuration, and a couple of others will still default to the in-tree approach. For the provider ready to be provisioned with the out-of-tree `CCM`, we could upgrade to this configuration and no longer support in-tree since the transition.

That's how I see this. Correct me if I'm missing something, but essentially if this ideally should be a painless approach.

Reply:
Can you clarify why Is this a non-goal? I think what we need is section on how we'd handle downgrades.

Reply:
Yes, starting with OpenStack in 4.7 and following up with other providers is in the plan. I can see no reason to move out to out-of-tree for AWS tomorrow, as the code there is just a straight import from the in-tree implementation - https://github.com/kubernetes/cloud-provider-aws/blob/master/cmd/aws-cloud-controller-manager/main.go#L57-L61

Reply:
Perhaps this can be reworded to make it more obvious what we mean by this? Does this suggestion capture the correct intent?
```suggestion
- Force an immediate transition for all cloud providers from in-tree support to their out-of-tree counterparts.
```

Reply:
Note that we need to validate each internal-to-external migration to make sure it's safe. eg, the last time I looked, the Azure out-of-tree provider was lagging several months behind the in-tree provider, so we'd regress people if we switched Azure to out-of-tree.